webapp/app/templates/lotse/form_steuernummer.html

```diff
@@ -0,0 +1,184 @@
+            }else if(taxNumberExistsNoField.checked && !taxNumberExistsNoField.wasCheckedBefore){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.StateSelectionNoTaxNumberExists;
+                taxNumberExistsYesField.wasCheckedBefore = false;
+                taxNumberExistsNoField.wasCheckedBefore = true;
+            }
+        }
+
+        function changeBundesland(){
+            let bundeslandField = document.getElementById('bundesland');
+            if (bundeslandField.value && taxNumberFormState === TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxNumberInput;
+            }else if(bundeslandField.value
+                     && (taxNumberFormState === TAX_NUMBER_FORM_STATES.StateSelectionNoTaxNumberExists
+                            || taxNumberFormState === TAX_NUMBER_FORM_STATES.TaxOfficeSelection)){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxOfficeSelection;
+                updateTaxOfficeList();
+            }else if(bundeslandField.value && taxNumberFormState === TAX_NUMBER_FORM_STATES.RequestNewTaxNumber){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxOfficeSelection;
+                updateTaxOfficeList();
+            }else if(!bundeslandField.value && taxNumberFormState === TAX_NUMBER_FORM_STATES.TaxNumberInput){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists;
+            }
+        }
+
+        function changeTaxOffice(){
+            if(document.getElementById('bufa_nr').value && taxNumberFormState === TAX_NUMBER_FORM_STATES.TaxOfficeSelection){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.RequestNewTaxNumber;
+            }else if(!document.getElementById('bufa_nr').value && taxNumberFormState === TAX_NUMBER_FORM_STATES.RequestNewTaxNumber){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxOfficeSelection;
+            }
+        }
+
+        function handleState(){
+            let stateDropdown = document.getElementById("bundesland");
+            let bufaNrDropdown = document.getElementById("bufa_nr");
+            let taxNumberInput = document.getElementById("steuernummer");
+            let requestNewTaxNumberCheckbox = document.getElementById("request_new_tax_number");
+            switch(taxNumberFormState){
+                case TAX_NUMBER_FORM_STATES.NothingSelected:
+                    deleteInput(stateDropdown);
+                    hideRequiredInput(stateDropdown);
+                    deleteInput(bufaNrDropdown);
+                    hideRequiredInput(bufaNrDropdown);
+                    deleteInput(taxNumberInput);
+                    hideRequiredInput(taxNumberInput);
+                    deleteInput(requestNewTaxNumberCheckbox);
+                    hideRequiredInput(requestNewTaxNumberCheckbox);
+                    break;
+                case TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists:
```

Comment:
Okay, this might be a provocative question, but why do we need the states at all? First I thought I liked it but then I saw this big switch-case and now I'm wondering: Why don't we just call these functions on e.g. dropdown change, so in the event handlers? I'm sure there is a great reason and I just don't see it yet.

Reply:
Imo there are two advantages of the states:

1. They decouple the logic and the visual changes. This makes it very easy to change how something should look like.

2. We can just go through all the different functions to check the current state and the change everything and delete data in every state. BTW: I could imagine that the flickering without the hidden class in Familienstand might happen because we change the state a few times.
This also decouples the states from each other

Reply:
I do really like the explicit state modelling, but I also think the boundaries don't sit in the ideal places yet. The giveaway for me is less the switch statement here (here we have a single, clearly-delineated code block for each sate) but more the many if-branches in `changeTaxNumberExists`/`changeBundesland`/`changeTaxOffice`, which I've found pretty hard to follow.

I'm not sure what a better approach would be, but since we're moving towards React, there's probably not too much point in working this out at this point…